### PR TITLE
Arithmetic

### PR DESCRIPTION
**Pull request by Bihaqo**
Implement add and multiply

### CODE REVIEW

File: initializers.py [hunk 1]
```diff
@@ -82,18 +81,17 @@ def random_matrix(shape, tt_rank=2):
...
     raise ValueError('`rank` array has inappropriate size')
 
   num_dims = shape[0].size
   if tt_rank.size == 1:
     tt_rank = tt_rank * np.ones(num_dims - 1)
-  # Add 1 to the beginning and end to simplify working with ranks.
-  tt_rank_ext = np.insert(tt_rank, 0, 1)
-  tt_rank_ext = np.append(tt_rank_ext, 1)
+    tt_rank = np.insert(tt_rank, 0, 1)
+    tt_rank = np.append(tt_rank, 1)
```
Comment (geffy) [inline initializers.py]: Maybe `tt_rank = np.concatenate([[1], tt_rank, [1]])`?

File: initializers.py [hunk 2]
```diff
@@ -82,18 +81,17 @@ def random_matrix(shape, tt_rank=2):
...
   # TODO: check that ints?
   shape = shape.astype(int)
-  tt_rank_ext = tt_rank_ext.astype(int)
+  tt_rank_ext = tt_rank.astype(int)
```
Comment (geffy) [inline initializers.py]: Why do we need `tt_rank_ext` in this function?

File: ops.py [hunk 3]
```diff
@@ -372,6 +372,153 @@ def flat_inner(a, b):
...
+    raise ValueError('The arguments should be both TT-tensors or both '
+                     'TT-matrices')
+
+  if tt_a.get_shape() != tt_b.get_shape():
+    raise ValueError('The arguments should have the same shape.')
+
+  static_a_ranks = tt_a.get_tt_ranks()
+  if static_a_ranks.is_fully_defined():
+    a_ranks = static_a_ranks.as_list()
+  else:
```
Comment (geffy) [inline ops.py]: Seems like it should be a function

File: ops.py [hunk 4]
```diff
@@ -372,6 +372,153 @@ def flat_inner(a, b):
...
+
+  # If tt_a.get_shape() is fully defined, it means that even in the TT-matrix
+  # case all dimensions are defined.
+  if tt_a.get_shape().is_fully_defined:
+    shape = [s.as_list() for s in tt_a.get_raw_shape()]
+  else:
+    shape = shapes.raw_shape(tt_a)
+
+  is_matrix = tt_a.is_tt_matrix()
+  if is_matrix:
```
Comment (geffy) [inline ops.py]: Maybe one-liner `last_axis= 3 if is_matrix else 2`?

File: ops.py [hunk 5]
```diff
@@ -372,6 +372,153 @@ def flat_inner(a, b):
...
+  tt_cores = []
+  for core_idx in range(ndims):
+    a_core = tt_a.tt_cores[core_idx]
+    b_core = tt_b.tt_cores[core_idx]
+    if core_idx == 0:
+      curr_core = tf.concat((a_core, b_core), axis=last_axis)
+    elif core_idx == ndims - 1:
+      curr_core = tf.concat((a_core, b_core), axis=0)
+    else:
+      if is_matrix:
```
Comment (geffy) [inline ops.py]: Maybe join this one and next `if`? (in order to make less segmented code)

File: ops.py [hunk 6]
```diff
@@ -372,6 +372,153 @@ def flat_inner(a, b):
...
+
+  Returns
+    a `TensorTrain` object corresponding to the element-wise product of the
+    arguments.
+
+  Raises
+    ValueError if the arguments shapes do not coincide.
+  """
+  ndims = tt_a.ndims()
+  if tt_a.is_tt_matrix() != tt_b.is_tt_matrix():
```
Comment (geffy) [inline ops.py]: initial shape and type checking should be a function

File: ops.py [hunk 7]
```diff
@@ -372,6 +372,153 @@ def flat_inner(a, b):
...
+      curr_core = tf.einsum('aijb,cijd->acijbd', a_core, b_core)
+      curr_core = tf.reshape(curr_core, (left_rank, shape[0][core_idx],
+                                         shape[1][core_idx], right_rank))
+    else:
+      curr_core = tf.einsum('aib,cid->acibd', a_core, b_core)
+      curr_core = tf.reshape(curr_core, (left_rank, shape[0][core_idx],
+                                         right_rank))
+    tt_cores.append(curr_core)
+
+  new_ranks = []
```
Comment (geffy) [inline ops.py]: Make as generator?

File: initializers.py (hunk 1 above)
Comment (Bihaqo) [inline initializers.py]: Done

File: initializers.py (hunk 2 above)
Comment (Bihaqo) [inline initializers.py]: Renamed to tt_rank

File: ops.py (hunk 5 above)
Comment (Bihaqo) [inline ops.py]: Doen

File: ops.py (hunk 3 above)
Comment (Bihaqo) [inline ops.py]: https://github.com/Bihaqo/t3f/issues/16

File: ops.py (hunk 6 above)
Comment (Bihaqo) [inline ops.py]: But the details are different in each scenario.. What do you think this function look like?

**Pull request merged**